starknet_devnet/blueprints/rpc.py

```diff
@@ -0,0 +1,681 @@
+    }
+
+
+async def get_block_transaction_count_by_hash(block_hash: str) -> int:
+    """
+    Get the number of transactions in a block given a block hash
+    """
+    try:
+        block = state.starknet_wrapper.blocks.get_by_hash(block_hash=block_hash)
+        return len(block.transactions)
+    except StarknetDevnetException as ex:
+        raise RpcError(code=24, message="Invalid block hash") from ex
+
+
+async def get_block_transaction_count_by_number(block_number: int) -> int:
+    """
+    Get the number of transactions in a block given a block number (height)
+    """
+    try:
+        block = state.starknet_wrapper.blocks.get_by_number(block_number=block_number)
+        return len(block.transactions)
+    except StarknetDevnetException as ex:
+        raise RpcError(code=26, message="Invalid block number") from ex
+
+
+async def call(contract_address: str, entry_point_selector: str, calldata: list, block_hash: str = "") -> list:
+    """
+    Call a starknet function without creating a StarkNet transaction
+    """
+    request_body = {
+        "contract_address": contract_address,
+        "entry_point_selector": entry_point_selector,
+        "calldata": calldata
+    }
+
+    # For now, we only support 'latest' block, support for specific blocks
+    # in devnet is more complicated if possible at all
+    if block_hash != "latest":
+        # By RPC here we should return `24 invalid block hash` but in this case I believe it's more
+        # descriptive to the user to use a custom error
+        raise RpcError(code=-1, message="Calls with block_hash != 'latest' are not supported currently.")
+
+    if not state.starknet_wrapper.contracts.is_deployed(int(contract_address, 16)):
+        raise RpcError(code=20, message="Contract not found")
+
+    try:
+        return await state.starknet_wrapper.call(transaction=make_invoke_function(request_body))
+    except StarknetDevnetException as ex:
+        raise RpcError(code=-1, message=ex.message) from ex
+    except StarkException as ex:
+        if f"Entry point {entry_point_selector} not found" in ex.message:
+            raise RpcError(code=21, message="Invalid message selector") from ex
+        if "While handling calldata" in ex.message:
+            raise RpcError(code=22, message="Invalid call data") from ex
+        raise RpcError(code=-1, message=ex.message) from ex
+
+
+async def get_block_number() -> int:
+    """
+    Get the most recent accepted block number
+    """
+    result = state.starknet_wrapper.blocks.get_number_of_blocks() - 1
+    return result if result >= 0 else 0
+
+
+async def chain_id() -> str:
+    """
+    Return the currently configured StarkNet chain id
+    """
+    chain: int = DEFAULT_GENERAL_CONFIG.chain_id.value
+    return hex(chain)
+
+
+async def pending_transactions():
+    """
+    Returns the transactions in the transaction pool, recognized by this sequencer
+    """
+    raise NotImplementedError()
```

Comment:
When I mentioned Not Implemented, I meant the 501 Not Implemented HTTP response

Reply:
Oh, I've misunderstood then. Perhaps I can add additional `except NotImplementedError` to base rpc route and then return a 501 http response. 

Another option would be to create return RPC error like all other errors with custom code (like `-2`) and a message.

Which one you think is better?

Reply:
I like the approach with a custom code (-2).

Reply:
Added custom error with -2